ykllvmwrap/src/ykllvmwrap.cc

```diff
@@ -113,19 +136,41 @@ extern "C" void *__ykllvmwrap_irtrace_compile(char *FuncNames[], size_t BBs[],
   if (TraceIdx == numeric_limits<uint64_t>::max())
     errx(EXIT_FAILURE, "trace index counter overflowed");
 
+  auto SrcMod = load_module(Context, SecPtr, SecSize);
+
+  // Get var args from start_tracing call.
+  auto Inputs = get_trace_inputs(SrcMod->getFunction(FuncNames[0]), BBs[0]);
+
+  std::vector<Type*> InputTypes;
+  for (auto Val: Inputs) {
+    InputTypes.push_back(Val->getType());
+  }
+
+  // Create function to store compiled trace.
   string TraceName = string(TRACE_FUNC_PREFIX) + to_string(TraceIdx);
   llvm::FunctionType *FType =
-      llvm::FunctionType::get(Type::getVoidTy(Context), false);
+      llvm::FunctionType::get(Type::getVoidTy(Context), InputTypes, false);
   llvm::Function *DstFunc = llvm::Function::Create(
       FType, Function::InternalLinkage, TraceName, DstMod);
   DstFunc->setCallingConv(CallingConv::C);
+
+  // Create entry block and setup builder.
   auto DstBB = BasicBlock::Create(Context, "", DstFunc);
   llvm::IRBuilder<> Builder(Context);
   Builder.SetInsertPoint(DstBB);
 
-  auto SrcMod = load_module(Context, SecPtr, SecSize);
-
   llvm::ValueToValueMapTy VMap;
+  // Map trace inputs to their appropriates variables in the trace.
```

Comment:
s/appropriates/appropriate/

Reply:
Also I'm not sure if "appropriate variables" is correct? Aren't they being mapped to function parameters?

Reply:
In the original code we have variables. When we construct the trace we then replace those variables with the trace inputs (which are passed as function arguments). I guess this could be reworded.

Reply:
Try bf1a44c.